smarts/env/gymnasium/hiway_env_v1.py

```diff
@@ -0,0 +1,390 @@
+            terminated (bool): Whether the agent reaches the terminal state (as defined under the MDP of the task)
+                which can be positive or negative. An example is reaching the goal state or moving into the lava from
+                the Sutton and Barton, Gridworld. If true, the user needs to call :meth:`reset`.
+            truncated (bool): Whether the truncation condition outside the scope of the MDP is satisfied.
+                Typically, this is a timelimit, but could also be used to indicate an agent physically going out of bounds.
+                Can be used to end the episode prematurely before a terminal state is reached.
+                If true, the user needs to call :meth:`reset`.
+            info (dict): Contains auxiliary diagnostic information (helpful for debugging, learning, and logging).
+                This might, for instance, contain: metrics that describe the agent's performance state, variables that are
+                hidden from observations, or individual reward terms that are combined to produce the total reward.
+                In OpenAI Gym <v26, it contains "TimeLimit.truncated" to distinguish truncation and termination,
+                however this is deprecated in favour of returning terminated and truncated variables.
+            done (bool): (Deprecated) A boolean value for if the episode has ended, in which case further :meth:`step` calls will
+                return undefined results. This was removed in OpenAI Gym v26 in favor of terminated and truncated attributes.
+                A done signal may be emitted for different reasons: Maybe the task underlying the environment was solved successfully,
+                a certain timelimit was exceeded, or the physics simulation has entered an invalid state.
+        """
+        assert isinstance(action, dict) and all(
+            isinstance(key, str) for key in action.keys()
+        ), "Expected Dict[str, any]"
+
+        observations, rewards, dones, extras = self._smarts.step(action)
+
+        infos = {
+            agent_id: {
+                "score": value,
+                "env_obs": observations[agent_id],
+                "done": dones[agent_id],
+                "reward": rewards[agent_id],
+            }
+            for agent_id, value in extras["scores"].items()
+        }
+
+        if self._env_renderer is not None:
+            self._env_renderer.step(observations, rewards, dones, infos)
+
+        for done in dones.values():
+            self._dones_registered += 1 if done else 0
+
+        dones["__all__"] = self._dones_registered >= len(self._agent_interfaces)
+
+        assert all("score" in v for v in infos.values())
+        return (
+            self.observations_formatter.format(observations),
+            sum(r for r in rewards.values()),
+            dones["__all__"],
+            dones["__all__"],
+            infos,
+        )
```

Comment:
Following RLlib, extending `terminated` and `truncated` to multi-agent case results in a dictionary.

https://github.com/ray-project/ray/blob/d909ac2b5fe5afe65feb192b49aca53abe962cee/rllib/env/multi_agent_env.py#L90-L132

Reply:
Right, but this is a gym environment rather than a `MultiAgentEnv` or `pettingzoo` environment. If we are going to do that we should first create a multi-agent interface or inherit from one of the two aforementioned environment types.

Since this is `gym` it should return gym environment level values rather than agent level values.

Reply:
The problem:
+ Gym and Gymnasium are only for single-agent.
+ Previously, we extended upon Gym to create a multi-agent SMARTS env `HiWayEnv`.
+ Previously, our multi-agent extension was in line with the multi-agent interface of other libraries such as RLlib.
+ RLlib also extends upon Gymnasium to create their multi-agent env `MultiAgentEnv`.
+ With the current changes, we have essentially extended upon Gymnasium to create a new multi-agent interface which is not in line with any other multi-agent libraries.
+ This is very likely to create more confusion.